src/main/java/org/example/bookapp/repository/SpecificationProviderManagerImpl.java

```diff
@@ -0,0 +1,23 @@
+package org.example.bookapp.repository;
+
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.example.bookapp.model.Book;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class SpecificationProviderManagerImpl implements SpecificationProviderManager<Book> {
+    private List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecification(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() ->
+                        new RuntimeException("Can't find correct specification provider for key "
+                                + key));
+    }
```

Comment:
```suggestion
                .filter(provider -> provider.getKey().equals(key))
                .findFirst()
                .orElseThrow(() ->
                        new NoSuchElementException("Can't find correct specification provider for key "
                                + key));
    }
```
it's better to throw more specific exceptions where possible, and avoid short names in stream chains